dns_sd_avahi.c

```diff
@@ -14,42 +15,191 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
+ * Some of this is insipred from libavahi's example:
+ * https://avahi.org/doxygen/html/client-browse-services_8c-example.html
+ * which is also LGPL 2.1 or later.
+ *
  * */
 
-#include "iio.h"
+#include "iio-private.h"
 
 #include <errno.h>
 #include <stdbool.h>
 #include <string.h>
 #include <time.h>
+#include <sys/time.h>
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
 
 #include <avahi-common/error.h>
 #include <avahi-common/simple-watch.h>
+#include <avahi-common/malloc.h>
 #include <avahi-client/client.h>
 #include <avahi-client/lookup.h>
 
 #include "debug.h"
 
+/* in network.c */
+int create_socket(const struct addrinfo *addrinfo, unsigned int timeout);
+#define DEFAULT_TIMEOUT_MS 5000
+#define IIOD_PORT 30431
+
+
 struct avahi_discovery_data {
 	AvahiSimplePoll *poll;
 	AvahiAddress *address;
-	uint16_t *port;
-	bool found, resolved;
+	char addr_str[AVAHI_ADDRESS_STR_MAX];
+	char *hostname;
+	uint16_t found, resolved;
+	uint16_t port;
+	struct avahi_discovery_data *next;
 };
 
+/*
+ * Fundamentally, this builds up a linked list to manage
+ * potential clients on the network
+ */
+
+static struct avahi_discovery_data * new_discovery_data(void)
```

Comment:
if this extends a linked-list, could we make this also add the new elements on the list?
and move some common stuff here?
maybe that could be among the first patches

Reply:
The current logic is different.

There is always a "blank" node ready for data to be populated in. 
See below, new_discovery_data is called/gets added to ->next when the current is populated. (as a blank one for next time). That is also how we find the end of the list in all the iterators.

Reply:
could we reverse the logic a bit and populate the node right after it gets created?

Reply:
Not really - it needs a blank node to store the pointer returned from `avahi_simple_poll_new();` in the `dnssd_find_hosts()` function - so it can be passed as part of the data structure (the linked list) to the browse and resolve function.

Reply:
oh, ok
makes sense